runtime/dancebox/src/lib.rs

```diff
@@ -811,54 +816,127 @@ impl RemoveInvulnerables<CollatorId> for RemoveInvulnerablesImpl {
-            let remaining_block_credits_to_pay = u128::from(remaining_block_credits).saturating_mul(block_production_costs);
-            let remaining_session_credits_to_pay = u128::from(remaining_session_credits).saturating_mul(collator_assignment_costs);
+        // If the para has been assigned collators for this session it must have enough block credits
+        // for the current and the next session.
+        let block_credits_needed = if currently_assigned.contains(&para_id) {
+            blocks_per_session * 2
+        } else {
+            blocks_per_session
+        };
+        // Check if the container chain has enough credits for producing blocks
+        let free_block_credits =
+            pallet_services_payment::BlockProductionCredits::<Runtime>::get(para_id)
+                .unwrap_or_default();
+        let remaining_block_credits = block_credits_needed.saturating_sub(free_block_credits);
+        let (block_production_costs, _) =
+            <Runtime as pallet_services_payment::Config>::ProvideBlockProductionCost::block_cost(
+                &para_id,
+            );
+        // Check if we can withdraw
+        let remaining_block_credits_to_pay =
+            u128::from(remaining_block_credits).saturating_mul(block_production_costs);
+        let remaining_to_pay = remaining_block_credits_to_pay;
+        // This should take into account whether we tank goes below ED
+        // The true refers to keepAlive
+        Balances::can_withdraw(
+            &pallet_services_payment::Pallet::<Runtime>::parachain_tank(para_id),
+            remaining_to_pay,
+        )
+        .into_result(true)?;
+        // TODO: Have proper weight
+        Ok(Weight::zero())
+    }
+}
 
-            let remaining_to_pay = remaining_block_credits_to_pay.saturating_add(remaining_session_credits_to_pay).saturating_add(max_tip);
+impl<AC> RemoveParaIdsWithNoCredits<BalanceOf<Runtime>, AC> for RemoveParaIdsWithNoCreditsImpl {
+    fn pre_assignment_remove_para_ids_with_no_credits(
+        para_ids: &mut Vec<ParaId>,
+        currently_assigned: &BTreeSet<ParaId>,
+    ) {
+        let blocks_per_session = Period::get();
+        para_ids.retain(|para_id| {
+            with_transaction(|| {
+                let max_tip =
+                    pallet_services_payment::MaxTip::<Runtime>::get(para_id).unwrap_or_default();
+                TransactionOutcome::Rollback(Self::charge_para_ids_internal(
+                    blocks_per_session,
+                    *para_id,
+                    currently_assigned,
+                    &Some(max_tip),
+                ))
+            })
+            .is_ok()
+        });
+    }
 
-            // This should take into account whether we tank goes below ED
-            // The true refers to keepAlive
-            Balances::can_withdraw(&pallet_services_payment::Pallet::<Runtime>::parachain_tank(*para_id), remaining_to_pay).into_result(true).is_ok()
+    fn post_assignment_remove_para_ids_with_no_credits(
```

Comment:
So can we remove this method? `pre_assignment` already uses `with_transaction` to remove para_ids that would have failed if `Self::charge_para_ids_internal` returns `Err`, so if I'm not wrong, calling `post_assignment` will never remove any additional para id?

Reply:
In perfect world yes. But, since we are changing state in between calling `pre` and `post` we need to run it again to eliminate even a tiniest possibility that some para_id escapes the restrictions.